### PR TITLE
Updated the logger to write to a CSV file and include the timestamp a…

### PR DESCRIPTION
**Pull request by luisdavidgarcia**
Updated the logger script with two CLI arguments one for log_dir and csv_file_name (name of the log file).

Also removed the print statements of the data since it is now being outputted to a CSV file.

### CODE REVIEW

File: src/Python/logger.py [hunk 1]
```diff
@@ -1,37 +1,68 @@
...
+
 # Open serial port
 try:
     ser = serial.Serial(args.port, args.baud_rate)
 except SerialException as ex:
     print(ex.strerror)
     exit()
 
-# Allow for Arduino to reset
+# Allow for Arduino to reset (physically press the reset button)
```
Comment (kennedyengineering) [inline src/Python/logger.py]: Not sure about Mac, but in my case the Arduino resets automatically when connected to over serial.

File: src/Python/logger.py [hunk 2]
```diff
@@ -1,37 +1,68 @@
...
 
 # Parse command line arguments
 parser = argparse.ArgumentParser()
 parser.add_argument("--port", type=str, default="/dev/ttyACM0")
 parser.add_argument("--baud_rate", type=int, default=115200)
+parser.add_argument("--log_dir", type=str, default="./logs")
+parser.add_argument("--csv_file_name", type=str, default=f"imu_data_{date_str}.csv")
 args = parser.parse_args()
 
+# Create log directory if it does not exist
```
Comment (kennedyengineering) [inline src/Python/logger.py]: Can ensure the log directory exists. Create empty logs/ and place a .gitignore inside of it with
```
*
!.gitignore
```

File: src/Python/logger.py [hunk 3]
```diff
@@ -1,37 +1,68 @@
...
 # Read data from Arduino
 try:
-    print("Starting logger.")
-    while True:
-        if ser.in_waiting > 0:  # Check if there is data to read
-            line = (
-                ser.readline().decode("utf-8").rstrip()
-            )  # Read a line from the serial
-            print(line)  # Print the line to the console
+    print(f"Starting logger. Writing to {file_path}")
```
Comment (kennedyengineering) [inline src/Python/logger.py]: Maybe include datetime in this print statement? Would be cool

File: src/Python/logger.py [hunk 4]
```diff
@@ -1,37 +1,68 @@
...
+    with open(file_path, "w", newline="") as csv_log_file:
+        writer = csv.writer(csv_log_file)
+        writer.writerow(
```
Comment (kennedyengineering) [inline src/Python/logger.py]: One thing that I am not particularly keen about is assuming knowledge of the headers. EX: timestamp, accelx, accely. What if code in tx changes?

File: src/Python/logger.py [hunk 5]
```diff
@@ -1,37 +1,68 @@
...
+                # IMU data is sent as a comma-separated string of key-value pairs
+                imu_data_pairs = line.split(",")
+                if len(imu_data_pairs) == 6:
+                    # Extract just the values from the key-value pairs
+                    parsed_imu_data = [
+                        float(pair.split(":")[1]) for pair in imu_data_pairs
+                    ]
+
+                    # Get the current timestamp
+                    timestamp = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
```
Comment (kennedyengineering) [inline src/Python/logger.py]: The type of timestamp needed here needs to go down to the microseconds, especially if a full 1khz MPU raw data sample rate is desired.

File: src/Python/logger.py [hunk 6]
```diff
@@ -1,37 +1,68 @@
...
+import os
+
+# Create timestamp to log entries and create log file
+date_str = time.strftime("%Y-%m-%d_%H-%M-%S", time.localtime())
 
 # Parse command line arguments
 parser = argparse.ArgumentParser()
 parser.add_argument("--port", type=str, default="/dev/ttyACM0")
 parser.add_argument("--baud_rate", type=int, default=115200)
+parser.add_argument("--log_dir", type=str, default="./logs")
```
Comment (kennedyengineering) [inline src/Python/logger.py:17]: I like the log dir idea. Not sure about the benefits of not being able to specify a filename though.

File: src/Python/logger.py (hunk 5 above)
Comment (kennedyengineering) [reply src/Python/logger.py]: This is technically a datetime that is used here. The date information will bloat the filesize of any large collection that takes place.

**Comment by kennedyengineering** [comment]
I like the direction this is headed.

It brings up a good question: when do you attach the timestamp? On the tx or rx side? This uses rx side, which is valid. My one concern would be if receiving stale data you wouldn't actually know when it was sampled. In other words, the assumption is that the time at which the data is received is approximately equal to the time at which it was sampled. An example of when this might not be the case is when the Arduino is connected to over serial -- it finishes its last tx, resets, and then continues to tx. On the rx side it can be difficult to differentiate the data from before and after the reset since the rx is buffered.

**Comment by kennedyengineering** [comment]
Sorry about the comments, the place I worked at was quite rigorous in their code reviews.

**Comment by kennedyengineering** [comment]
In my ideal world, I would have two tasks setup in the MCU:

1. On MCU interrupt; read data via I2C, get MCU timestamp, and place into a queue.
2. Main task; pop from queue, tx data over serial.

In this case testing would need to be done to see if an overflow occurs in the tx queue. If so, adjustments to sample frequency would have to be made. (Like queuing data in every other interrupt, etc.)

However it might be simpler to modify the code to include the timestamp in the collection of print statements it currently has. See     https://github.com/kennedyengineering/EE525_Final_Project/blob/a8ea29e6a6ddec4aeefceb3ba539bf6ca5566b69/src/Arduino/main/main.ino#L37

**Comment by kennedyengineering** [comment]
From my experience in dealing with post processing data, accurate timing information is everything.

File: src/Python/logger.py (hunk 1 above)
Comment (luisdavidgarcia) [reply src/Python/logger.py]: For Mac it's needed sadly, well it's odd if I use "screen" don't need to reset, but with the script I need to

**Comment by kennedyengineering** [comment]
Made what I think is the easiest change. See MR #2

**Comment by kennedyengineering** [comment]
Since screen can work as a Linux/OSX solution, maybe we don't reinvent the wheel with this logger.py program? What if we created a src/Scripts directory and created a bash script that did the logging using screen? We could save the src/Python directory for post-processing scripts maybe...

**Comment by kennedyengineering** [comment]
Synthesized ideas into an issue #3

**Pull request closed without merging**